ledger/simulation/simulation_eval_test.go

```diff
@@ -2190,3 +2226,387 @@ func TestMockTracerScenarios(t *testing.T) {
+								Txns: []simulation.TxnResult{
+									{
+										Txn: transactions.SignedTxnWithAD{
+											ApplyData: transactions.ApplyData{
+												EvalDelta: transactions.EvalDelta{
+													InnerTxns: make([]transactions.SignedTxnWithAD, innerCount),
+												},
+											},
+										},
+										AppBudgetConsumed: ignoreAppBudgetConsumed,
+									},
+								},
+								AppBudgetAdded:    700 + 700*uint64(innerCount),
+								AppBudgetConsumed: ignoreAppBudgetConsumed,
+							},
+						},
+						EvalOverrides: simulation.ResultEvalOverrides{
+							AllowUnlimitedResourceAccess: true,
+						},
+					},
+				}
+			})
+		})
+	}
+}
+
+// TestUnlimitedResourceAccessAccountLocalWrite tests app call behavior when writing to an account's
+// local state they otherwise shouldn't have access to if AllowUnlimitedResourceAccess is enabled.
+func TestUnlimitedResourceAccessAccountLocalWrite(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+	// Start with directRefEnabledVersion (4), since prior to that all restricted references had to
+	// be indexes into the foreign arrays, meaning we can't test the "unlimited" case.
+	for v := 4; v <= logic.LogicVersion; v++ {
+		v := v
+		t.Run(fmt.Sprintf("v%d", v), func(t *testing.T) {
+			t.Parallel()
+			simulationTest(t, func(env simulationtesting.Environment) simulationTestCase {
+				sender := env.Accounts[0]
+				testAppUser := env.Accounts[6].Addr
+
+				program := fmt.Sprintf(`#pragma version %d
+txn ApplicationID
+!
+txn OnCompletion
+int OptIn
+==
+||
+bnz end // Do nothing during create or opt in
+
+// App local write to an account we shouldn't be able to
+addr %s
+byte "key"
+byte "value"
+app_local_put
+
+end:
+int 1
+`, v, testAppUser)
+
+				testAppID := env.CreateApp(sender.Addr, simulationtesting.AppParams{
+					ApprovalProgram:   program,
+					ClearStateProgram: fmt.Sprintf("#pragma version %d\n int 1", v),
+					LocalStateSchema: basics.StateSchema{
+						NumByteSlice: 1,
+					},
+				})
+
+				env.OptIntoApp(testAppUser, testAppID)
+
+				txn := env.TxnInfo.NewTxn(txntest.Txn{
+					Type:          protocol.ApplicationCallTx,
+					Sender:        sender.Addr,
+					ApplicationID: testAppID,
+				})
+				stxn := txn.Txn().Sign(sender.Sk)
+
+				var expectedEvalDelta transactions.EvalDelta
+				var expectedError string
+				var expectedFailedAt simulation.TxnPath
+				// Can write to accounts outside of foreign array in sharedResourcesVersion (9), but not before then
+				if v >= 9 {
+					expectedEvalDelta = transactions.EvalDelta{
+						SharedAccts: []basics.Address{testAppUser},
+						LocalDeltas: map[uint64]basics.StateDelta{
+							1: {
+								"key": basics.ValueDelta{
+									Action: basics.SetBytesAction,
+									Bytes:  "value",
+								},
+							},
+						},
+					}
+				} else {
+					expectedError = fmt.Sprintf("logic eval error: invalid Account reference for mutation %s", testAppUser)
+					expectedFailedAt = simulation.TxnPath{0}
```

Comment:
this error is triggered in `eval.go :: mutableAccountReference`, from `opAppLocalPut`.